dora/core/common/src/main/java/alluxio/membership/ServiceDiscoveryRecipe.java

```diff
@@ -0,0 +1,323 @@
+import io.etcd.jetcd.op.Cmp;
+import io.etcd.jetcd.op.CmpTarget;
+import io.etcd.jetcd.op.Op;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.PutOption;
+import io.etcd.jetcd.support.CloseableClient;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * ServiceDiscoveryRecipe for etcd, to track health status
+ * of all registered services.
+ */
+public class ServiceDiscoveryRecipe {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final String BASE_PATH = "/ServiceDiscovery";
+  AlluxioEtcdClient mAlluxioEtcdClient;
+  ScheduledExecutorService mExecutor;
+  String mClusterIdentifier = "";
+  // Will look like /ServiceDiscovery/<mClusterIdentifier>
+  String mRegisterPathPrefix = "";
+  @SuppressFBWarnings({"URF_UNREAD_FIELD"})
+  private final ReentrantLock mRegisterLock = new ReentrantLock();
+  final ConcurrentHashMap<String, ServiceEntity> mRegisteredServices = new ConcurrentHashMap<>();
+
+  /**
+   * CTOR for ServiceDiscoveryRecipe.
+   * @param client
+   * @param clusterIdentifier
+   */
+  public ServiceDiscoveryRecipe(AlluxioEtcdClient client, String clusterIdentifier) {
+    mAlluxioEtcdClient = client;
+    mAlluxioEtcdClient.connect();
+    mClusterIdentifier = clusterIdentifier;
+    mRegisterPathPrefix = String.format("%s%s%s", BASE_PATH,
+        MembershipManager.PATH_SEPARATOR, mClusterIdentifier);
+    mExecutor = Executors.newSingleThreadScheduledExecutor(
+        ThreadFactoryUtils.build("service-discovery-checker", false));
+    mExecutor.scheduleWithFixedDelay(this::checkAllForReconnect,
+        AlluxioEtcdClient.DEFAULT_LEASE_TTL_IN_SEC, AlluxioEtcdClient.DEFAULT_LEASE_TTL_IN_SEC,
+        TimeUnit.SECONDS);
+  }
+
+  /**
+   * Apply for a new lease for given ServiceEntity.
+   * @param service
+   * @throws IOException
+   */
+  private void newLeaseInternal(ServiceEntity service) throws IOException {
+    try (LockResource lockResource = new LockResource(service.mLock)) {
+      if (service.mLease != null && !mAlluxioEtcdClient.isLeaseExpired(service.mLease)) {
+        LOG.info("Lease attached with service:{} is not expired, bail from here.");
```

Comment:
I don't quite understand the content of this log msg. Maybe make it more clear to the user?

Reply:
this call is shared by create new lease (start first heartbeat) and lease expired trying to create new lease ( heartbeat for some reason cut off trying to restart the heartbeat) , so this log is for the 2nd case, if we got false alarm that heartbeat got cut off, so when we reach here to check lease hasnt expired, we will bail from here instead of creating new lease